aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -71,42 +119,52 @@ Tensor to(
   bool copy,
   c10::optional<c10::MemoryFormat> optional_memory_format
 ) {
-  TensorOptions options = TensorOptions()
-      .dtype(dtype)
-      .layout(layout)
-      .device(device)
-      .pinned_memory(pin_memory)
-      .memory_format(optional_memory_format);
-
-  TORCH_CHECK(!options.has_layout() || self.layout() == options.layout(),
-           "to(options) doesn't support converting to a different layout, "
-           "but got self.layout being ", self.layout(),
-           " and options.layout set as ", options.layout());
-
-  if (options.has_device()) {
-    options = options.device(ensure_has_index(options.device()));
-  }
-  auto specified_options = self.options().merge_in(options);
-  return to_impl(self, specified_options, non_blocking, copy);
+  return to_impl(
+      self,
+      dtype,
+      layout,
+      device.has_value() ? ensure_has_index(*device) : device,
```

Comment:
nit: make ensure_has_index overload for optional?

Reply:
Aha, didn't realize it had one. Thanks!